src/include/txn.i

```diff
@@ -769,9 +775,49 @@ __wt_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_UPDATE *upd, WT
         if (upd_visible == WT_VISIBLE_PREPARE)
             return (WT_PREPARE_CONFLICT);
     }
+    WT_ASSERT(session, upd == NULL);
+
+    /*
+     * FIXME-PM-1521: In some code paths, we don't expect to access the ondisk value. If we're doing
```

Comment:
This is my main issue right now. In some invocations of `__wt_txn_read`, we were never falling back to the on-disk value. If I try to access the on-disk value, I hit issues because `cbt->slot` is purposely set to `UINT32_MAX`. In the cases where we weren't falling back to the on-disk value, I'm supplying `skip_disk=true` as a temporary solution to mimic what we were doing before. However, I don't see a reason why it would be ok to just ignore the on-disk value in any of the code paths so I think the existing search behaviour is wrong.

The code where we're setting `cbt->slot` to a bad value is in `bt_curnext.c:__cursor_row_next` (and the prev counterparts) where we hit the end of a page and then move to the next one. I'm not sure why moving to a new page means we can't read the on-disk value and also the general mechanics around how the `cbt->slot` works are hard for me to wrap my head around.

@keithbostic Could you shed some light on this?

Reply:
What's the stack on failure?

I doubt seriously you're hitting the values set by next/prev.

I'm wondering if you're running into this code [here](https://github.com/wiredtiger/wiredtiger/blob/wt-5271-remove-page-las/src/reconcile/rec_row.c#L604-L612) and [here](https://github.com/wiredtiger/wiredtiger/blob/wt-5271-remove-page-las/src/reconcile/rec_col.c#L1045-L1050).

Reply:
This one is readily reproducible if you remove the `goto` that I've used to get around this problem.
```
alexc@alexc:~/work/wiredtiger/build_posix$ LD_LIBRARY_PATH=.libs lldb python ../test/suite/run.py cursor_random
(lldb) target create "python"
Current executable set to 'python' (x86_64).
(lldb) settings set -- target.run-args  "../test/suite/run.py" "cursor_random"
(lldb) r
Process 31909 launched: '/usr/bin/python' (x86_64)
....ssssssss..................................Process 31909 stopped
* thread #1, name = 'python', stop reason = signal SIGSEGV: invalid address (fault address: 0x7fffffff8)
    frame #0: 0x00007ffff5790e64 libwiredtiger-3.2.2.so`__wt_row_leaf_value(page=0x000055555606ab00, rip=0x00000007fffffff8, value=0x00007fffffffc938) at btree.i:1038
   1035	    uintptr_t v;
   1036
   1037	    /* The row-store key can change underfoot; explicitly take a copy. */
-> 1038	    v = (uintptr_t)WT_ROW_KEY_COPY(rip);
   1039
   1040	    /*
   1041	     * See the comment in __wt_row_leaf_key_info for an explanation of the magic.
(lldb) bt
* thread #1, name = 'python', stop reason = signal SIGSEGV: invalid address (fault address: 0x7fffffff8)
  * frame #0: 0x00007ffff5790e64 libwiredtiger-3.2.2.so`__wt_row_leaf_value(page=0x000055555606ab00, rip=0x00000007fffffff8, value=0x00007fffffffc938) at btree.i:1038
    frame #1: 0x00007ffff5790c73 libwiredtiger-3.2.2.so`__wt_value_return_buf(cbt=0x0000555555e6a360, ref=0x0000555555cfe8b0, buf=0x00007fffffffc938, start=0x00007fffffffc928, stop=0x00007fffffffc918) at bt_ret.c:107
    frame #2: 0x00007ffff5754fa7 libwiredtiger-3.2.2.so`__wt_txn_read(session=0x0000555555dacf40, cbt=0x0000555555e6a360, upd=0x0000000000000000, updp=0x00007fffffffca28, skip_disk=true) at txn.i:789
    frame #3: 0x00007ffff57539b3 libwiredtiger-3.2.2.so`__cursor_row_prev(cbt=0x0000555555e6a360, newpage=true, restart=false) at bt_curprev.c:509
    frame #4: 0x00007ffff57525be libwiredtiger-3.2.2.so`__wt_btcur_prev(cbt=0x0000555555e6a360, truncating=false) at bt_curprev.c:638
    frame #5: 0x00007ffff5785100 libwiredtiger-3.2.2.so`__random_leaf(cbt=0x0000555555e6a360) at bt_random.c:329
    frame #6: 0x00007ffff5784b40 libwiredtiger-3.2.2.so`__wt_btcur_next_random(cbt=0x0000555555e6a360) at bt_random.c:522
    frame #7: 0x00007ffff581e131 libwiredtiger-3.2.2.so`__wt_curfile_next_random(cursor=0x0000555555e6a360) at cur_file.c:118
    frame #8: 0x00007ffff5bd2d2d _wiredtiger.so`_wrap_Cursor_next(self=<unavailable>, args=<unavailable>) at wiredtiger_wrap.c:4161
    frame #9: 0x0000555555651c85 python`PyEval_EvalFrameEx + 35573
    frame #10: 0x000055555564e452 python`PyEval_EvalFrameEx + 21186
    frame #11: 0x0000555555646d3a python`PyEval_EvalCodeEx + 1402
    frame #12: 0x0000555555662c6c python`___lldb_unnamed_symbol3285$$python + 748
    frame #13: 0x0000555555632a1e python`PyObject_Call + 62
    frame #14: 0x000055555564b7ec python`PyEval_EvalFrameEx + 9820
    frame #15: 0x0000555555646d3a python`PyEval_EvalCodeEx + 1402
    frame #16: 0x00005555556629c9 python`___lldb_unnamed_symbol3285$$python + 73
    frame #17: 0x000055555567b3de python`___lldb_unnamed_symbol3398$$python + 414
    frame #18: 0x0000555555632a1e python`PyObject_Call + 62
    frame #19: 0x00005555556d8db3 python`___lldb_unnamed_symbol3970$$python + 51
    frame #20: 0x000055555564e640 python`PyEval_EvalFrameEx + 21680
    frame #21: 0x0000555555646d3a python`PyEval_EvalCodeEx + 1402
    frame #22: 0x0000555555662c6c python`___lldb_unnamed_symbol3285$$python + 748
    frame #23: 0x0000555555632a1e python`PyObject_Call + 62
    frame #24: 0x000055555564b7ec python`PyEval_EvalFrameEx + 9820
    frame #25: 0x0000555555646d3a python`PyEval_EvalCodeEx + 1402
    frame #26: 0x00005555556629c9 python`___lldb_unnamed_symbol3285$$python + 73
    frame #27: 0x000055555567b3de python`___lldb_unnamed_symbol3398$$python + 414
    frame #28: 0x0000555555632a1e python`PyObject_Call + 62
    frame #29: 0x00005555556d8db3 python`___lldb_unnamed_symbol3970$$python + 51
    frame #30: 0x000055555564e640 python`PyEval_EvalFrameEx + 21680
    frame #31: 0x000055555564e452 python`PyEval_EvalFrameEx + 21186
    frame #32: 0x0000555555646d3a python`PyEval_EvalCodeEx + 1402
    frame #33: 0x000055555564ee63 python`PyEval_EvalFrameEx + 23763
    frame #34: 0x0000555555646d3a python`PyEval_EvalCodeEx + 1402
    frame #35: 0x00005555556467b9 python`PyEval_EvalCode + 25
    frame #36: 0x00005555556774df python`___lldb_unnamed_symbol3378$$python + 47
    frame #37: 0x00005555556722d2 python`PyRun_FileExFlags + 130
    frame #38: 0x000055555567174d python`PyRun_SimpleFileExFlags + 397
    frame #39: 0x0000555555620de4 python`Py_Main + 1620
    frame #40: 0x00007ffff7a05b97 libc.so.6`__libc_start_main(main=(python`main), argc=3, argv=0x00007fffffffe478, init=<unavailable>, fini=<unavailable>, rtld_fini=<unavailable>, stack_end=0x00007fffffffe468) at libc-start.c:310
    frame #41: 0x000055555562069a python`_start + 42
(lldb) f 3
frame #3: 0x00007ffff57539b3 libwiredtiger-3.2.2.so`__cursor_row_prev(cbt=0x0000555555e6a360, newpage=true, restart=false) at bt_curprev.c:509
   506 	        if ((ins = cbt->ins) != NULL) {
   507 	            key->data = WT_INSERT_KEY(ins);
   508 	            key->size = WT_INSERT_KEY_SIZE(ins);
-> 509 	            WT_RET(__wt_txn_read(session, cbt, ins->upd, &upd, true));
   510 	            if (upd == NULL)
   511 	                continue;
   512 	            if (upd->type == WT_UPDATE_TOMBSTONE) {
(lldb) p newpage
(bool) $0 = true
(lldb) p cbt->slot
(uint32_t) $1 = 4294967295
```

Reply:
Here's a trivial script that triggers the failure:
```
import unittest, wiredtiger, wttest
from wtdataset import simple_key, simple_value
class test_xxx(wttest.WiredTigerTestCase):
    def test_xxx(self):
        uri = 'file:xxx'
        self.session.create(uri, 'key_format=S,value_format=S')
        self.session.begin_transaction()
        cursor = self.session.open_cursor(uri, None)
        for i in range(1, 5):
            cursor[simple_key(cursor, i)] = simple_value(cursor, i)
            
        # Open another session, the updates won't yet be visible, we shouldn't
        # find anything at all. 
        s = self.conn.open_session()
        cursor = s.open_cursor(uri, None)
        self.assertEqual(cursor.next(), wiredtiger.WT_NOTFOUND)
        
if __name__ == '__main__':
    wttest.run()
```
If you set a breakpoint and look at the underlying page, you get:
```
Thread 1 "python" hit Breakpoint 1, __wt_txn_read (session=0x555555dbd7b0,
cbt=0x555555e65370, upd=0x0, updp=0x7fffffffcf40, 
    skip_disk=true) at ./src/include/txn.i:786
786	    if (skip_disk)
(gdb) print __wt_debug_page(session, 0, cbt.ref, "/dev/stdout")
0x555555e64ed0: row-store leaf
	disk (nil), entries 0, dirty, page-state=2, memory-size 1064
	insert {000000000000001}
	value {1: abcdefghijklmnopqrstuvwxyz}
	txn id 3, start_ts (0, 0)
	insert {000000000000002}
	value {2: abcdefghijklmnopqrstuvwxyz}
	txn id 3, start_ts (0, 0)
	insert {000000000000003}
	value {3: abcdefghijklmnopqrstuvwxyz}
	txn id 3, start_ts (0, 0)
	insert {000000000000004}
	value {4: abcdefghijklmnopqrstuvwxyz}
	txn id 3, start_ts (0, 0)
$2 = 0
(gdb) p *cbt.ref.page
$5 = {
  u = {
    intl = {
      parent_ref = 0x0, 
      split_gen = 0, 
      __index = 0x0
    }, 
    row = 0x0, 
    fix_bitf = 0x0, 
    col_var = {
      col_var = 0x0, 
      repeats = 0x0
    }
  }, 
  entries = 0, 
  type = 7 '\a', 
  flags_atomic = 0 '\000', 
  unused = "\000", 
  read_gen = 201, 
  memory_footprint = 1064, 
  dsk = 0x0,                                    <<<< No disk image
  modify = 0x555555e80190, 
  cache_create_gen = 1, 
  evict_pass_gen = 0
}
```
In other words, there's no underlying disk image, so there's no on-page key/value pairs to select from.

Reply:
Thanks for taking a look. My bad, I got distracted by the `newpage` value and didn't notice that. It makes sense for me to guard that `__wt_value_return_buf` call so I'll add this check shortly.